docs/proposals/unified-backend-architecture/tech-spec.md

```diff
@@ -0,0 +1,270 @@
+sign their transactions and send it for publishing at the same time, the node must be able to handle these
+transactions reliably and as fast as possible.
+
+To achieve this goal the node must implement robust concurrent model. More specifically, all requests to the wallet
+web API endpoints must be handled concurrently. In addition, it is possible to use queues/buffers, it will allow
+to keep sustained transactions submission rate.
+
+In case of correct design decisions, even the slowest endpoints, like `/api/v1/transactions`, can be handled fully
+concurrently (publishing to the network, storing in the `wallet-db`, etc).
+
+_Alfredo said: AFAIK that's possible, `acid-state` supports atomic transactions and all the rest of it. Provided we
+write the right `Update` transactions we should be fine._
+
+## Scalability
+
+Initially clusters of nodes will be hosted by IOHK (on AWS EC2). And we need a mechanism of auto-scaling of the
+cluster: it must be able to grow proportionally to growing number of clients.
+
+**TODO**: To think about auto-scale supervisor. Suppose we have 1 cluster that consists of 100 nodes and each of node
+can reliably serve 100 clients. If supervisor sees that number of clients connected to this cluster is growing (and
+becomes more than 10000), it launches additional nodes. But exact resource and tuning to do to be decided during
+later benchmarking.
+
+_Domen said: We might not base our autoscaling on number of users, but another resource. It's hard to say how it
+would scale before we run some benchmarks. It could be that bottleneck is CPU and you can't correlate that with number
+of subscribers._
+
+**TODO 2**: Launching additional node is relatively expensive action (in terms of time and memory). So it is better
+if one single node will be able to serve more clients than just increase a number of nodes in the cluster.
+We have to understand what the real bottlnecks are. E.g. for relays, the number of clients syncing concurrently
+_is_ the bottleneck, not the number of client just keeping up with the blockchain.
+
+## User Stories
+
+These are typical user stories (workflows), with description of API calls.
+
+### Initial State
+
+After installation of mobile client application user does not see any wallets, and he must create at least one to
+work with a client application.
+
+During the first connection with the node the new session is establishing, please read an explanation below.
+
+### Wallet Creation
+
+User creates a new wallet. It produces a `POST`-request to `/api/v1/wallets/external` (with information about client's
+public key), node creates a new wallet in `wallet-db` and sends response with information about new wallet.
+
+Even if node `N` which serves the client `C` will die, another node `M` (the client `C` will reconnect to) will be able
+to handle information about `C`'s wallet(s), because `wallet-db` is storing in the cluster.
```

Comment:
Ok, let's reason about this one. We are a user and we want to create a wallet. In a normal scenario we would ask Daedalus to do this for us, feeding as input the 12 words mnemonic. Then calling `/api/v1/wallets` would give us back a new `Wallet` object, effectively creating it in the backend.

But in the context of the unified backend architecture (UBA) we do not want to share _any_ sensible information with the external world, so we won't send to `N` our 12 words mnemonic but just the _hash_ of the root public key for this HD wallet. In a nutshell, we would do client side what we are now doing on the server side. Let's note a couple of things:

* Incidentally, this was the original idea we had when we thought about the integration with external hardware wallets. If you take a look at [this endpoint](https://rebilly.github.io/ReDoc/?url=https://gist.githubusercontent.com/adinapoli-iohk/93791a963a9eb545315840c784fab3c2/raw/bbf320febe511839ce0545d669463fc13aee1742/stub.yaml#/paths/~1api~1wallets~1external/post), it pretty much accepts such _hash_ as input.

* If we were using random index derivation it wouldn't be safe to share such _hash_, as currently this is used (AFAIK) as a piece of secret for some symmetric encryption used to encrypt the random index into each `AddrAttributes` (I think that's called `HDPayload`).

Please @darko-mijic and @vincenthz tell me if I'm mistaken here. If I'm right, I would extend this user story to be a bit more precise and less vague 😁 

🚴 shedding: Let's come back to the choice of the name for the endpoint, as part of CHW-37.